18/README.md

```diff
@@ -0,0 +1,119 @@
+
+Of greatest interest, we can see that in a series of tests where we separate broadcasts with a pause of varying duration, the transaction rate is stabilized. For example, examine lines 126 to 194 of the [data]. In this series of tests, we send 10 request batches. Each batch lasts 100 seconds, and has a request rate of 1,000 requests per second. We vary the pauses between request batch from 1 to 20 seconds in successive tests, and in all cases, our transaction rate stabilizes between 900 and 1,000 transactions per second. Compare this to sending 1,000 requests per second for 1,000 seconds straight, as we did on line 13 of the [data]. Without pauses, our average transaction rate fell to 210 transactions per second.
+
+Note that there is a sweet spot here. When we reduce our broadcast time too much, we start to lose performance, (see lines 254 - 294 in the [data]). This is due to the fact that there is a lag when we initialize the connection to Tendermint. There is a cost to breaking and re-establishing this connection that needs to be balanced against the increase in transaction rate.
+
+## Recommendations
+
+We recommend implementation of a sentinel server to field incoming requests to Tendermint. This will offer several benefits.
+
+First, we can use the sentinel to protect against DoS attacks, by limiting the number of requests from any given IP address to a reasonable number.
+
+                                            _____________________                                ____________
+    User_1, 100,000 requests/s------------>| Sentinel            |   User_1, 995 requests/s     |            |
+    User_2, 1,000 requests/s-------------->| max_req=2,000/s     |---User_2, 995 requests/s---> | Tendermint |
+    User_3, 10 requests/s----------------->| pause_interval=100s |   User_3, 10 requests/s      |____________|
+                                           |_____________________|   Break every 100s
+
+Second, we can use the sentinel to break and reinitialize connection with Tendermint as a mitigation for the perfomance degradation.
+
+This could be implemented in two ways. First, we could simply experiment to find the optimal frequency with which we hang up and call back to Tendermint. This has the advantage of being easy to engineer. However, we could also implement active monitoring from the sentinel layer, where we measure the transaction rate and hang up if it falls below a certain threshold. This offers the advantage of being a more general solution that will respond to changes in performance as different version of Tendermint are rolled out, or when the system is run on different hardware, for example.
```

Comment:
Might make sense to mention [AIMD](https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease) here.

Reply:
I don't think rate limiting the txs will help. I see the same performance decay even when I run at 10 requests per second.

Reply:
I am not talking about rate limiting. On the contrary, we can allegedly maintain a high external rate limit by varying the pressure we put internally on Tendermint. For that, we would need a buffer in between.

Reply:
Yes. My idea here was that the sentinel server would remain open to external requests all the time, and store excess requests in a buffer. It would manage the forwarding of requests to TM without breaking connections with the outside world.